### PR TITLE
Add function mapContainsKeyValue and answer it from the keyValuePairs text index

### PR DESCRIPTION
**Pull request by Ergus**
`mapContainsKeyValue(map, key, value)` tells whether a map holds an entry with
both that key and that value, equivalent to
`arrayExists((k, v) -> k = key AND v = value, mapKeys(map), mapValues(map))`.
Unlike `map[key] = value`, it considers every entry, so it also matches a
repeated key whose later occurrence carries the value.

A text index with the `keyValuePairs` tokenizer answers it exactly: the pair has
one token for a first occurrence of the key and one for a repetition, so the
index searches both and direct read removes the predicate. `map[key] = ''` still
falls back to a scan, but `mapContainsKeyValue(map, key, '')` does not, because
an empty value is a stored pair like any other.

Related: https://github.com/ClickHouse/ClickHouse/pull/115579

### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Added function `mapContainsKeyValue(map, key, value)`, which returns whether a map contains an entry with the given key and value. In contrast to `map[key] = value`, all entries of a repeated key are considered. A text index with the `keyValuePairs` tokenizer evaluates the function directly from the index.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=120244&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/120244](https://github.com/search?q=head%3Async-upstream%2Fpr%2F120244+org%3AClickHouse+type%3Apr&type=pullrequests)]

### CODE REVIEW

File: src/Functions/map.cpp [hunk 1]
```diff
@@ -397,6 +399,172 @@ class FunctionMapUpdate final : public IFunction
...
+        else
+            needle = needle_argument.column->replicate(offsets);
+        needle = recursiveRemoveLowCardinality(needle);
+
+        const bool elements_are_null = isAlwaysNullOrEmpty(element_type);
+        const bool needle_is_null = isAlwaysNullOrEmpty(needle_type);
+        const NullMap * element_nulls = getNullMap(*elements);
+        const NullMap * needle_nulls = getNullMap(*needle);
+
+        ColumnPtr equals_result;
```
Comment (lukaszszafranski) [inline src/Functions/map.cpp:538]: _🎯 Correctness_ · _🟡 Medium_

`getNullMap()` only recognizes a direct `ColumnNullable`, but `needle` remains a `ColumnConst` after `cloneResized()`. Consequently a typed constant nullable needle such as `CAST(NULL AS Nullable(String))` has no detected null map, so a NULL map value is treated as not matching even though the documented contract says NULL matches NULL. Materialize or unwrap const needles before reading nullability, or use a null check that handles `ColumnConst`, and add a typed-NULL regression.

<details>
<summary>🔍 Evidence</summary>

Bare `NULL` is special-cased through `Nullable(Nothing)`, but `Nullable(String)` is not always-null; its null bit remains inside the const column.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this against the current code and fix it only if it still applies; keep the change minimal and validate.

In @src/Functions/map.cpp around line 527, make typed constant NULL needles participate in the same null-map path as materialized nullable needles and add a focused regression.
```

</details>

File: src/Functions/map.cpp [hunk 2]
```diff
@@ -457,6 +625,34 @@ For two maps, returns the first map with values updated on the values for the co
     FunctionDocumentation::Category category_mapUpdate = FunctionDocumentation::Category::Map;
     FunctionDocumentation documentation_mapUpdate = {description_mapUpdate, syntax_mapUpdate, arguments_mapUpdate, {}, returned_value_mapUpdate, examples_mapUpdate, introduced_in_mapUpdate, category_mapUpdate};
     factory.registerFunction<FunctionMapUpdate>(documentation_mapUpdate);
+
+    /// mapContainsKeyValue function documentation
+    FunctionDocumentation::Description description_mapContainsKeyValue = R"(
+Returns whether the map contains an entry with the given key and value, i.e. whether
+`arrayExists((k, v) -> k = key AND v = value, mapKeys(map), mapValues(map))` holds.
```
Comment (lukaszszafranski) [inline src/Functions/map.cpp]: _🩺 Stability_ · _🔵 Low_

Qualify this equivalence to non-NULL keys and values. The implementation explicitly makes NULL equal NULL, whereas the displayed expression uses ordinary `=` comparisons, for which a NULL operand produces NULL rather than true. As written, the public documentation gives two contradictory contracts for nullable maps.

<details>
<summary>🔍 Evidence</summary>

The following paragraph says that a NULL matches only another NULL, and the tests expect `1` for a stored NULL searched with NULL.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this against the current code and fix it only if it still applies; keep the change minimal and validate.

In @src/Functions/map.cpp around line 632, qualify the ordinary-equality expression to non-NULL inputs or rewrite it to express the implemented null-safe behavior.
```

</details>

**Review by lukaszszafranski — COMMENTED** [review]
Thanks @alexey-milovidov for the opportunity to run our tool on this PR. Here are the findings that we found:

## Veridical.dev reviewed the change:

- **Merge readiness** — ⛔ Do not merge
- **Correctness** — 🟠 Needs work
- **Security** — No security defect identified; not a security assurance
- **Tests** — 🟡 The typed-NULL path needs a focused regression
- **Changed hunks with no finding** — 74% (finding distribution only, not review recall)

**3 surfaced items · 2 with source/evidence text attached**

<details>
<summary>📝 Walkthrough</summary>

Adds `mapContainsKeyValue` with repeated-key and NULL-aware semantics, registers and documents it, and lets `keyValuePairs` text indexes answer the predicate through exact direct reads.

| File | Change |
| :-- | :-- |
| `src/Functions/map.cpp` | Implements, validates, documents, and registers `mapContainsKeyValue`. |
| `src/Storages/MergeTree/MergeTreeIndexConditionText.cpp` | Recognizes the function and searches first and repeated pair tokens. |
| `src/Storages/MergeTree/registerStorageMergeTree.cpp` | Adds the function to the text-index support matrix. |
| `docs/reference/engines/table-engines/mergetree-family/textindexes.mdx` | Documents indexed support and limitations. |
| Four stateless test/reference files | Cover function semantics, comparisons, edge cases, and index behavior. |

</details>

<details>
<summary>🗺️ How the changes interact</summary>

```mermaid
flowchart TD
A[mapContainsKeyValue registration] --> B[Map function]
B --> C[NULL-aware matching]
D[keyValuePairs text index] --> E[Index condition traversal]
E --> F[Exact direct read]
B --> G[Function tests]
F --> H[Index tests]
I[Documentation] --> B
I --> F
```

</details>

### Supported defects

| Severity | Where | Issue |
| :-- | :-- | :-- |
| 🟡 Medium | [`src/Functions/map.cpp:527`](https://github.com/ClickHouse/ClickHouse/blob/3e88817caf590a3bdeda0e0c497b09167c722b72/src/Functions/map.cpp#L527) | `getNullMap()` sees only a direct `ColumnNullable`, while a typed constant NULL remains wrapped in `ColumnConst`. |
| 🔵 Low | [`src/Functions/map.cpp:632`](https://github.com/ClickHouse/ClickHouse/blob/3e88817caf590a3bdeda0e0c497b09167c722b72/src/Functions/map.cpp#L632) | The documented ordinary-equality expression contradicts the function’s stated NULL-equals-NULL behavior. |

### Nonblocking observations

- `matchEntries` allocates a second `num_entries` buffer and copies equality flags in the non-nullable path. This is an optimization hypothesis only; no material performance impact was established.

### Findings that were reported by other bots

None on the exact reviewed head. The live discussion contains one CI report, its Code Review section is empty, and there are no GitHub review comments.

<details>
<summary>🔬 How the two supported defects were verified</summary>

- **`src/Functions/map.cpp:527`** — bare `NULL` uses `Nullable(Nothing)`, but `Nullable(String)` is not always-null; after `cloneResized`, its null bit remains hidden inside `ColumnConst` and `getNullMap()` returns null.
- **`src/Functions/map.cpp:632`** — the next paragraph promises that NULL matches NULL, and the tests expect that result, while the displayed expression uses ordinary `=` comparisons.

</details>

---

### Review scope: complete

The reviewed revision changes **9 files / 592 added lines**. All nine changed files were supplied to the review; no changed file was skipped.

---

<sub>Reviewed by <b>Veridical</b> · ultra depth · <i>some optional execution checks did not complete</i></sub>

File: src/Storages/MergeTree/MergeTreeIndexConditionText.cpp [hunk 3]
```diff
@@ -1938,6 +1946,47 @@ bool MergeTreeIndexConditionText::traverseMapElementKeyValueNode(
...
+    if (!hasIndexForColumn(function_node.getArgumentAt(0).getColumnName()))
+        return false;
+
+    /// FixedString excluded: its padding is compared away by the function but not stored in the token,
+    /// so exact direct read would drop matching rows. Same as traverseMapElementKeyValueNode.
+    auto get_string_constant = [](const RPNBuilderTreeNode & node) -> std::optional<String>
+    {
+        Field field;
+        DataTypePtr type;
+        if (!node.tryGetConstant(field, type) || !WhichDataType(type).isString())
```
Comment (diegomestre2) [inline src/Storages/MergeTree/MergeTreeIndexConditionText.cpp:1964]: what happens with  m['key'] = toLowCardinality('v') in this case here?

**Review by diegomestre2 — APPROVED** [review]
The solution is very clear and straightforward. I just left one question regarding a specific case but the overall code is in good shape to be merged once CI is all green!
